src/System.Private.CoreLib/shared/Internal/Runtime/CompilerServices/Unsafe.cs

```diff
@@ -357,6 +357,17 @@ public static void Write<T>(ref byte destination, T value)
             return ref Unsafe.As<byte, T>(ref *(byte*)source);
         }
 
+        /// <summary>
+        /// Reinterprets the given location as a reference to a value of type <typeparamref name="T"/>.
+        /// </summary>
+        [Intrinsic]
+        [NonVersionable]
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static ref T AsRef<T>(in T source)
```

Comment:
I understand this is in the `Unsafe` class hence it's unsafe. I want to emphasize though that this operation is not just about violating `readonly` semantics. It actually can cause some pretty severe safety issues. Consider the following code as an example:

``` c#
void Example(in object o) { 
  Unsafe.AsRef(in o) = new object();
}

void M(string[] array) { 
  object[] objArray = array;
  Example(in objArray[0]);
}
```

Hence this operator allows writing objects that violate our basic type safety into locations. It's not just about violating `readonly`.

Reply:
Right. S.R.CS.Unsafe is a very sharp knife that lets introduce pretty severe safety issue if used incorrectly. This particular API shipped already. This change is just adding internal clone of the public API for internal CoreLib consumption.